### DISCUSSION TITLE
Should I Expend the Effort?

### DISCUSSION THREAD

**Discussion by johnmarksilly** [blakeblackshear/frigate#3359] [Polls]
# Realtime Events and QuickClips™ Playback

As events are triggered, the UI receives them in real-time using WebSockets.
I have a separate process running to pick up events after the clips are generated and convert them to shortened clips.

Video below:
[![frigate clips demo](https://img.youtube.com/vi/jnXdJ3bnonU/0.jpg)](https://www.youtube.com/watch?v=jnXdJ3bnonU)

## What am I proposing?

The poll is about the video playback, not the UI. I would be integrating the playback and realtime events in the clips page.

**Comment by NickM-27** [comment]
Speaking strictly subjectively, I'd find this most useful as like a mouse-over feature on the current events page snapshots (or maybe for all visible events without more over). I would find it less useful taking up its own space without the other timing / label metadata. 

Regardless, very cool and I do think many would find it useful in some form

**Reply by johnmarksilly** [reply]
Yeah I should have prefaced that the functionality is the clips, not the design haha. I decided to build my own UI because I don't like the way it exists now.

Basically if I was to re-implement it in Frigate, it would be in the events page (obviously more refined, but the idea):

https://user-images.githubusercontent.com/11224731/174349511-6fab2065-92c3-4900-bc02-0f1f82d6ff5a.mov

**Reply by NickM-27** [reply]
Ah yes, exactly what I was thinking 🎉

Also, not sure how you're implementing it, but if it were implemented as an API endpoint then it could probably be utilized for notifications as well.

**Comment by blakeblackshear** [comment]
There have been discussions about generating these kinds of thumbnail clips in a few places. This should be done as a general purpose post processor that could be used to perform other things like convert from h265 to h264. I think it would be best to generate a single video file based on the segments in a day or hour and reuse it for the event previews by seeking to the correct time. We wouldn't want to generate a bunch of these for individual events.

**Reply by johnmarksilly** [reply]
Yeah agreed. I currently set it up as a general purpose post processor actually. I'm making a video to go over how it works. I plan on doing full hour clips 1 frame per second or one frame per minute picking frames that have interesting content so scrubbing through full length recordings process faster. I actually really like the processing of each event. The file sizes are seriously small. They are smaller than the actual event thumbnails.

**Reply by johnmarksilly** [reply]
I'm always open to chatting over the wire anytime as well. I'm really interested in building out a lot of other integrations by using events.

**Comment by johnmarksilly** [comment]
I posted a video on how it works. If I was to implement it in Frigate I would use MQTT without Kafka entirely.

https://www.youtube.com/watch?v=jnXdJ3bnonU

**Reply by asantaga** [reply]
Very nice, I like the impl of Kafka ( I use it at work ) but as you say , to keep things simple in frigate Id probably do it directly in mqtt..

**Reply by johnmarksilly** [reply]
I use Kafka at work as well and already had a cluster running on my servers for other projects. Being able to have messages stored on disk and replay them helps with development.

**Comment by mitchross** [comment]
Where is the code for this?

**Comment by mitchross** [comment]
@johnmarksilly  any update?

**Comment by johnmarksilly** [comment]
So here is the first iteration: https://github.com/brilliant-monkey/frigate-clips

The above repo contains all the details needed to add the `clips` project which will generate video thumbnails once an event ends and publish it over a `frigate/clips` MQTT topic.

This can be picked up by Home Assistant to forward the video to end devices via Push Notifications.

Enjoy.

**Reply by mitchross** [reply]
Can you make a branch with the kafka version?

**Reply by johnmarksilly** [reply]
Probably not as I removed all of the code for the Kafka version. There were more people integrating with MQTT and Home Assistant and I figured I would follow that pattern as that's what Frigate supports. 

If you want to push to Kafka, I'm sure you could find a proxy tool out there to re-publish events from MQTT to Kafka.

**Reply by johnmarksilly** [reply]
Here is something ~~https://dzone.com/articles/streaming-iot-data-and-mqtt-messages-to-apache-kaf~~ https://stackoverflow.com/questions/33374422/using-mqtt-protocol-with-kafka-as-a-message-broker
